sled-agent/config-reconciler/src/internal_disks.rs

```diff
@@ -386,28 +494,332 @@ impl cmp::PartialOrd for InternalDiskDetailsId {
+    fn from(disk: Disk) -> Self {
+        // Invariant: We should only be constructed for internal disks; our
+        // caller should have already filtered out all external disks.
+        match disk.variant() {
+            DiskVariant::M2 => (),
+            DiskVariant::U2 => panic!("InternalDisk called with external Disk"),
+        }
+        Self { identity: Arc::new(disk.identity().clone()), disk }
+    }
+}
+
+impl Deref for InternalDisk {
+    type Target = Disk;
+
+    fn deref(&self) -> &Self::Target {
+        &self.disk
+    }
+}
+
+impl IdMappable for InternalDisk {
+    type Id = Arc<DiskIdentity>;
+
+    fn id(&self) -> Self::Id {
+        Arc::clone(&self.identity)
+    }
+}
+
 struct InternalDisksTask {
-    disks_tx: watch::Sender<Arc<IdMap<InternalDiskDetails>>>,
+    // Input channel for changes to the raw disks sled-agent sees.
     raw_disks_rx: watch::Receiver<IdMap<RawDiskWithId>>,
-    mount_config: Arc<MountConfig>,
 
-    // Invokes dumpadm(8) and savecore(8) when new disks are encountered
-    dump_setup: DumpSetup,
+    // The set of disks we've successfully started managing.
+    disks_tx: watch::Sender<IdMap<InternalDisk>>,
 
+    // Output channel summarizing any adoption errors.
+    //
+    // Because this is a different channel from `disks_tx`, it isn't possible
+    // for a caller to get an atomic view of "the current disks and errors", so
+    // it's possible a disk could appear in both or neither if snapshots are
+    // taken at just the wrong time. This isn't great, but is maybe better than
+    // squishing both into a single channel. If we combined them, then we'd wake
+    // up any receivers that really only care about disk changes any time we had
+    // changes to errors. (And if we get an error, we'll retry; if it continues
+    // to fail, we'd continue to wake up receivers after each retry, because we
+    // don't have an easy way of checking whether errors are equal.)
+    //
+    // The errors are only reported in inventory, which already has similar
+    // non-atomic properties across other fields all reported together.
+    errors_tx: watch::Sender<Arc<BTreeMap<DiskIdentity, DiskError>>>,
+
+    mount_config: Arc<MountConfig>,
     log: Logger,
 }
 
 impl InternalDisksTask {
-    async fn run(self) {
-        unimplemented!()
+    async fn run<T: DiskAdopter>(mut self, disk_adopter: T) {
+        // If disk adoption fails, the most likely cause is that the disk is not
+        // formatted correctly, and we have no automated means to recover that.
+        // However, it's always possible we could fail to adopt due to some
+        // transient error. Construct an exponential backoff that scales up to
+        // waiting a minute between attempts; that should let us handle any
+        // short transient errors without constantly retrying a doomed
+        // operation.
+        //
+        // We could be smarter here and check for retryable vs non-retryable
```

Comment:
I think it's reasonable to set a limit here, and just report the disk as failed after a few attempts. However, as long as we can accurately get access to the errors and push them up to a higher level, then it probably doesn't matter. 

I'm not even sure we can accurately tell the difference between a transient and permanent failure, so this seems fine for now.